File: Leviathan/Source/Mathematics/Vec3.cs
using System.Numerics;

namespace Leviathan.Mathematics
{
	public struct Vec3
	{
		public float Magnitude => MathF.Sqrt(x * x + y * y + z * z);
		
		public float x = 0;
		public float y = 0;
		public float z = 0;

		public Vec3(float _x, float _y, float _z)
		{
			x = _x;
			y = _y;
			z = _z;
		}

		//Dot
		public float Dot(Vec3 _rhs) => x * _rhs.x + y * _rhs.y + z * _rhs.z;

		public Vec3 Cross(Vec3 _rhs)
		{
			return new Vec3(y * _rhs.z - z * _rhs.y,
				z * _rhs.x - x * _rhs.z,
				x * _rhs.y - y * _rhs.x);
		}

		//Normalisation - Non Static
		public void Normalise()
		{
			float length = Magnitude;
			if(length == 0)
			{
				x = 0;
				y = 0;
				z = 0;
			}
			else
			{
				x /= length;
				y /= length;
				z /= length;
			}
		}
		
		public bool Equals(Vec3 _other) => x.Equals(_other.x) && y.Equals(_other.y) && z.Equals(_other.z);

		public override bool Equals(object? _obj) => _obj is Vec3 other && Equals(other);

		public override int GetHashCode() => HashCode.Combine(x, y, z);

		public override string ToString() => $"({x}, {y}, {z})";

		//Normalisation - Static
		public static Vec3 Normalise(Vec3 _vec)
		{
			float len = _vec.Magnitude;
			return len == 0 ? new Vec3(0, 0, 0) : new Vec3(_vec.x / len, _vec.y / len, _vec.z / len);
		}

		public static float Dot(Vec3 _lhs, Vec3 _rhs) => _lhs.x * _rhs.x + _lhs.y * _rhs.y + _lhs.z * _rhs.z;

		/// <summary>
		/// This method allows an implicit conversion from System.Numerics.Vector2 to our Vec3 class.
		/// </summary>
		/// <param name="_vec"></param>
		public static implicit operator Vec3(Vector3 _vec) => new Vec3(_vec.X, _vec.Y, _vec.Z);

		/// <summary>
		/// This method allows an implicit conversion from our Vec3 class to System.Numerics.Vector2.
		/// </summary>
		/// <param name="_v"></param>
		public static implicit operator Vector3(Vec3 _v) => new Vector3(_v.x, _v.y, _v.z);

		public static bool operator ==(Vec3 _lhs, Vec3 _rhs) => _lhs.Magnitude - _lhs.Magnitude < Math.EPSILON && _lhs.Magnitude - _rhs.Magnitude > -Math.EPSILON;
		
		public static bool operator !=(Vec3 _lhs, Vec3 _rhs) => !(_lhs == _rhs);

		//Addition Operator Overload +
		public static Vec3 operator +(Vec3 _lhs, Vec3 _rhs) => new Vec3(_lhs.x + _rhs.x, _lhs.y + _rhs.y, _lhs.z + _rhs.z);

		//Subtraction Operator Overload - 
		public static Vec3 operator -(Vec3 _lhs, Vec3 _rhs) => new Vec3(_lhs.x - _rhs.x, _lhs.y - _rhs.y, _lhs.z - _rhs.z);

		//PreScale Operator Overload 
		public static Vec3 operator *(float _lhs, Vec3 _rhs) => new Vec3(_lhs * _rhs.x, _lhs * _rhs.y, _lhs * _rhs.z);

		//PostScale Operator Overload
		public static Vec3 operator *(Vec3 _lhs, float _rhs) => new Vec3(_lhs.x * _rhs, _lhs.y * _rhs, _lhs.z * _rhs);
	}
}

File: Framework/Leviamath/vec3.cpp
#include <Leviathan/Math/vec3.h>

#include <Leviathan/Math/leviamath.h>

vec3::vec3() : x(0), y(0), z(0)
{
}

vec3::vec3(float _val) : x(_val), y(_val), z(_val)
{
}

vec3::vec3(float _x, float _y, float _z)
	: x(_x), y(_y), z(_z)
{
}

vec3::vec3(initializer_list<float> _values) : x(0), y(0), z(0)
{
	int i = 0;
	for (const float* iter = _values.begin(); iter != _values.end(); iter++)
	{
		values[i] = *iter;
		i++;
	}
}

float vec3::magnitude()
{
	return leviamath::sqrt(sqrMagnitude());
}

float vec3::sqrMagnitude()
{
	return x * x + y * y + z * z;
}

vec3 vec3::normalised()
{
	float mag = magnitude();
	vec3 normal = vec3(0);

	if (mag != 0)
	{
		normal.x = x / mag;
		normal.y = y / mag;
		normal.z = z / mag;
	}

	return normal;
}

void vec3::normalise()
{
	float mag = magnitude();

	if (mag != 0)
	{
		x /= mag;
		y /= mag;
		z /= mag;
	}
	else
	{
		x = y = z = 0;
	}
}

vec3 vec3::normalise(vec3 _vec)
{
	float mag = _vec.magnitude();
	vec3 normal = _vec;

	if (mag != 0)
	{
		normal.x = _vec.x / mag;
		normal.y = _vec.y / mag;
		normal.z = _vec.z / mag;
	}

	return normal;
}

float vec3::dot(vec3 _lhs, vec3 _rhs)
{
	return _lhs.x * _rhs.x + _lhs.y * _rhs.y + _lhs.z * _rhs.z;
}

vec3 vec3::cross(vec3 _lhs, vec3 _rhs)
{
	return vec3(
		_lhs.y * _rhs.z - _lhs.z * _rhs.y,
		_lhs.z * _rhs.x - _lhs.x * _rhs.z,
		_lhs.x * _rhs.y - _lhs.y * _rhs.x);
}
